packages/cubejs-backend-shared/src/time.ts

```diff
@@ -26,10 +31,113 @@ export const TIME_SERIES: Record<string, (range: DateRange, timestampPrecision:
     .map(d => [d.format(`YYYY-MM-DDT00:00:00.${'0'.repeat(digits)}`), d.endOf('quarter').format(`YYYY-MM-DDT23:59:59.${'9'.repeat(digits)}`)]),
 };
 
-type TimeSeriesOptions = {
-  timestampPrecision: number
+/**
+ * Parse PostgreSQL-like interval string into object
+ * E.g. '2 years 15 months 100 weeks 99 hours 15 seconds'
+ * Negative units are also supported
+ * E.g. '-2 months 5 days -10 hours'
+ */
+export function parseSqlInterval(intervalStr: SqlInterval): ParsedInterval {
+  const interval: ParsedInterval = {};
+  const parts = intervalStr.split(/\s+/);
+
+  for (let i = 0; i < parts.length; i += 2) {
+    const value = parseInt(parts[i], 10);
+    const unit = parts[i + 1];
+
+    // Remove ending 's' (e.g., 'days' -> 'day')
+    const singularUnit = (unit.endsWith('s') ? unit.slice(0, -1) : unit) as unitOfTime.DurationConstructor;
+    interval[singularUnit] = value;
+  }
+
+  return interval;
+}
+
+export function addInterval(date: moment.Moment, interval: ParsedInterval): moment.Moment {
+  const res = date.clone();
+
+  Object.entries(interval).forEach(([key, value]) => {
+    res.add(value, key as unitOfTime.DurationConstructor);
+  });
+
+  return res;
+}
+
+export function subtractInterval(date: moment.Moment, interval: ParsedInterval): moment.Moment {
+  const res = date.clone();
+
+  Object.entries(interval).forEach(([key, value]) => {
+    res.subtract(value, key as unitOfTime.DurationConstructor);
+  });
+
+  return res;
+}
+
+/**
+ * Returns the closest date prior to date parameter aligned with the origin point
+ */
+function alignToOrigin(startDate: moment.Moment, interval: ParsedInterval, origin: moment.Moment): moment.Moment {
+  let alignedDate = startDate.clone();
+  let intervalOp;
+  let isIntervalNegative = false;
+
+  let offsetDate = addInterval(origin, interval);
+
+  // The easiest way to check the interval sign
+  if (offsetDate.isBefore(origin)) {
+    isIntervalNegative = true;
+  }
+
+  offsetDate = origin.clone();
+
+  if (startDate.isBefore(origin)) {
+    intervalOp = isIntervalNegative ? addInterval : subtractInterval;
+
+    while (offsetDate.isAfter(startDate)) {
```

Comment:
I am thinking about protection for super large time series. for example, user can specify 2 seconds interval from 1970 to 2050 and it will do a CPU spike (I don't think that manipulation with dates are fast + OOM?)

I belive, we should care about it. It's vector for DDoS